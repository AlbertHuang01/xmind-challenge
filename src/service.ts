import { Bill, BillCategory } from './model';
import axios from 'axios';

const API = {
  BILL_LIST: '/bill/list',
  CATEGORY_LIST: '/category/list',
}

export const queryBillList = async (): Promise<Bill[]> => {
  const resp = await axios.get(API.BILL_LIST,)
  return resp.data
}

export const queryCategoryList = async (): Promise<BillCategory[]> => {
  const resp = await axios.get(API.CATEGORY_LIST,)
  return resp.data
}
