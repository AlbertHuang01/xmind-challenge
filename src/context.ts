import { queryBillList, addBillItem, queryCategoryList } from './service';
import { Bill, BILL_TYPE, BillCategory } from './model';
import { createContext, useEffect, useMemo, useState, useContext } from "react";
import dayjs from 'dayjs';

interface Condition {
  date?: {
    year: number;
    month: number;
  },
  categoryId?: string
}

const context = createContext({})

export const Provicer = context.Provider

export function initContext() {
  const [billList, setBillList] = useState<Bill[]>([])
  const [categoryList, setCategoryList] = useState<BillCategory[]>([])
  const [condition, setCondition] = useState<Condition>(null)
  const [addBillVisible,setAddBillVisible]=useState(false)

  const billListFilted = useMemo(() => {
    let result = billList;
    if (condition) {
      const { date, categoryId } = condition
      if (date) {
        const { year, month } = date
        result = result.filter(bill => {
          const time = dayjs(bill.time)
          return time.year() === year && time.month() + 1 === month
        })
      }
      if (categoryId) {
        result = result.filter(bill => bill.category === categoryId)
      }
    }
    return result
  }, [billList, condition])

  const billListGroupByType = useMemo(() => {
    const income = billListFilted.filter(bill => bill.type === BILL_TYPE.INCOME).reduce((prev, curr) => prev + curr.amount, 0)
    const expenditure = billListFilted.filter(bill => bill.type === BILL_TYPE.EXPENDITURE).reduce((prev, curr) => prev + curr.amount, 0)
    return { income, expenditure }
  }, [billListFilted])

  const addBill = (bill: Bill) => {
    addBillItem(bill).then(() => {
      queryBillList().then(setBillList)
    })
  }

  useEffect(() => {
    queryBillList().then(setBillList)
    queryCategoryList().then(setCategoryList)
  }, [])

  return {
    billList: billListFilted,
    billListGroupByType,
    condition,
    setCondition,
    addBill,
    categoryList,
    addBillVisible,
    setAddBillVisible
  }
}

export const useAppContext = () => {
  return useContext(context) as ReturnType<typeof initContext>
}
