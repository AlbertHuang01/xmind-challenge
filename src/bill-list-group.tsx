import { useContext } from "react";
import { APP_DATA } from "./main";
import { BILL_TYPE } from "./model";
import { Typography } from "antd";

const { Title, Paragraph } = Typography;

export function GroupByIncomeAndExpenditure() {
  const { billList } = useContext(APP_DATA);

  const income = billList
    .filter((bill) => bill.type === BILL_TYPE.INCOME)
    .reduce((pre, curr) => pre + curr.amount, 0);
  const expenditure = billList
    .filter((bill) => bill.type === BILL_TYPE.EXPENDITURE)
    .reduce((pre, curr) => pre + curr.amount, 0);

  return (
    <Typography>
      <Title>收入支出分统计</Title>
      <Paragraph>收入：{income} ￥</Paragraph>
      <Paragraph>支出：{expenditure} ￥</Paragraph>
    </Typography>
  );
}

export function GroupByCategoryType() {
  const { billList, categories } = useContext(APP_DATA);
  const group = categories
    .map((category) => {
      const foundBills = billList.filter((bill) => bill.category === category.id);
      const totalAmount = foundBills.reduce((pre, curr) => pre + curr.amount, 0);
      return {
        name: category.name,
        totalAmount,
      };
    })
    .sort((a, b) => a.totalAmount - b.totalAmount);

  return (
    <Typography>
      <Title>账单分类分组统计</Title>
      <Paragraph>
        <ul>
          {group.map((item) => (
            <li key={item.name}>
              {item.name}：{item.totalAmount}￥
            </li>
          ))}
        </ul>
      </Paragraph>
    </Typography>
  );
}
