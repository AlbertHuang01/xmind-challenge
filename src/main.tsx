import { useEffect } from "react";
import axios from "axios";

import SearchForm from "./search-form";
import zhCN from "antd/lib/locale/zh_CN";
import { Card, ConfigProvider, Space } from "antd";
import "./main.css";
import BillList from "./bill-list";
import AddBill from "./add-bill";
import { GroupByCategoryType, GroupByIncomeAndExpenditure } from "./bill-list-group";
import { appContextInit, APP_DATA_CONTEXT } from "./context";

// 设置 axios 请求时的 baseURL
axios.defaults.baseURL = "http://127.0.0.1:3000";

export const APP_DATA = APP_DATA_CONTEXT

export function App() {
  const appContext=appContextInit()

  return (
    <ConfigProvider locale={zhCN}>
      <Card style={{ width: "80%", margin: "40px auto" }}>
        <APP_DATA.Provider
          value={appContext}
        >
          <Space direction={"vertical"} style={{ display: "flex" }}>
            <SearchForm />
            <BillList />
          </Space>
          <AddBill />
          <GroupByIncomeAndExpenditure />
          <GroupByCategoryType />
        </APP_DATA.Provider>
      </Card>
    </ConfigProvider>
  );
}
