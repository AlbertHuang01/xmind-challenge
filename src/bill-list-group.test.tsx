import { render } from "@testing-library/react"
import { renderHook } from "@testing-library/react-hooks"
import { act } from "react-dom/test-utils"
import { GroupByCategoryType, GroupByIncomeAndExpenditure } from "./bill-list-group"
import { appContextInit, useAppContext } from "./context"
import { APP_DATA } from "./main"
import { getBillList } from "./service"

describe('bill-list-group',()=>{
  it('GroupByIncomeAndExpenditure snapshot',async()=>{
    const {waitForNextUpdate,result}=renderHook(()=>appContextInit())
    await waitForNextUpdate()

    const wrapper = ({ children }) => <APP_DATA.Provider value={result.current}>{children}</APP_DATA.Provider>

    const {baseElement}=render(<GroupByIncomeAndExpenditure/>,{wrapper})
    expect(baseElement).toMatchSnapshot()
  })
  // it('GroupByCategoryType snapshot',()=>{
  //   const {baseElement}=render(<GroupByCategoryType/>)
  //   expect(baseElement).toMatchSnapshot()
  // })
})
