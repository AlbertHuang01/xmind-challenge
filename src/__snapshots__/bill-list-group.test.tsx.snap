// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bill-list-group GroupByIncomeAndExpenditure snapshot 1`] = `
<body>
  <div>
    <article
      class="ant-typography"
    >
      <h3
        class="ant-typography"
      >
        收入支出分组统计
      </h3>
      <div
        class="ant-typography"
      >
        收入：
        134001
         ￥
      </div>
      <div
        class="ant-typography"
      >
        支出：
        104800
         ￥
      </div>
    </article>
  </div>
</body>
`;
